CorsixTH/Lua/rooms/ward.lua

```diff
@@ -123,36 +124,54 @@ function WardRoom:doStaffUseCycle(humanoid)
   humanoid:queueAction(MeanderAction():setLoopCallback(meanders_loop))
 end
 
+function WardRoom:calculateHealingFactor()
+  local patient_count = 1
+  local nurse_factor = 0
+  for humanoid in pairs(self.humanoids) do
+    if not humanoid:isLeaving() then
+      if class.is(humanoid, Patient) then
+        patient_count = patient_count + 1
+      elseif humanoid.humanoid_class == "Nurse" and not humanoid.fired then
+        nurse_factor = nurse_factor + 0.5 + humanoid:getServiceQuality()
+      end
+    end
+  end
+  -- Difficulty of healing - 10% faster on easier, 10% slower on hard mode
+  local difficulty_factor = 0.8 + self.world.map:getDifficulty() * 0.1
+  if nurse_factor > 0 then
+    self.healing_amount = nurse_factor / math.log(patient_count) / difficulty_factor
+  else
+    self.healing_amount = 0
+  end
+end
+
 
--- TODO the nurse should not leave the ward if there are beds in use, therefore prevent her from being picked up
--- and have a system that stops patients entering the ward if she is in need of taking a break or being called elsewhere.
 function WardRoom:commandEnteringPatient(patient)
-  local staff = next(self.staff_member_set) or self.staff_member
   local bed, pat_x, pat_y = self.world:findFreeObjectNearToUse(patient, "bed")
   self:setStaffMembersAttribute("dealing_with_patient", nil)
   if not bed then
     patient:setNextAction(self:createLeaveAction())
     patient:queueAction(self:createEnterAction(patient))
     print("Warning: A patient was called into the ward even though there are no free beds.")
-  else
-    bed.reserved_for = patient
-    self:countWorkingNurses()
-    local length = (math.random(200, 800) * (1.5 - staff.profile.skill))  / self.nursecount -- reduce time in ward if there is more than one nurse on duty
-    local --[[persistable:ward_loop_callback]] function loop_callback(action)
-    -- TODO Perhaps it should take longer if there are more used beds!
-      if length <= 0 then
-        action.prolonged_usage = false
-      end
-      length = length - 1
-    end
-    local after_use = --[[persistable:ward_after_use]] function()
-      self:dealtWithPatient(patient)
+    return
+  end
+
+  bed.reserved_for = patient
+  local healing_required = math.random(150, 600)
+  local --[[persistable:ward_loop_callback]] function loop_callback(action)
+    healing_required = healing_required - self.healing_amount
```

Comment:
I think this will break on loading games that have `ward_loop_callback` in the queue because healing_required won't be defined - changing upvalues is tricky.

Reply:
Can confirm this breaks, but looks like Toby has dealt with it now

```
A stack trace is included below, and the handler has been disconnected.
Lua\rooms\ward.lua:143: use of undeclared variable 'healing_required'
stack traceback:
        [C]: in function 'error'
        Lua\strict.lua:43: in metamethod '__index'
        Lua\rooms\ward.lua:143: in method 'loop_callback'
        Lua\humanoid_actions\use_object.lua:324: in local 'timer_function'
        Lua\entity.lua:216: in method 'tick'
        Lua\world.lua:1072: in method 'onTick'
        Lua\app.lua:1128: in function <Lua\app.lua:1125>
        (...tail calls...)
        Lua\app.lua:1028: in function <Lua\app.lua:1023>
```

Reply:
Fixed, if I understand it right. It works for me.